### PR TITLE
[Messenger] Fix exception message of failed message is dropped on retry

### LINKED ISSUE #32311: [Messenger] Messages retried on failure transport lose exception metadata
**Symfony version(s) affected**: 4.3.*

**How to reproduce**  
* Try to handle a messenger message that fails... and allow it to fail until it goes to the failure transport
* Once on the failure transport, use `messenger:failed:retry` to retry the message. But, make sure it fails *again*.
* Try to view the specific information about the message - e.g. `messenger:failed:show <id>`. It will lose details about its exception. I think they are probably still there... but not being read correctly.

Message on failure transport before retrying:
before

Message on failure transport after retrying:
after

**Possible Solution**  
We need to figure out if the exception information is being lost on redelivery (and if that's the case, make it not lost) or if it is present, but not being read correctly.

### LINKED ISSUE #32719: [Messenger] Message retried in failure transport dropped on exception
**Symfony version(s) affected**: 4.3.2

**Description**  
Message retried in failure transport dropped on exception

**How to reproduce**  
I configured a failure transport on symfony messenger (like explain here: https://symfony.com/blog/new-in-symfony-4-3-messenger-failure-transport)
I throw an error on my handler, the message is retried 3 times, and then sent to the failure transport.

Then, I retry the failed message with `bin/console messenger:consume failed` or `bin/console messenger:failed:retry {id}`.
The message handling fail again and the message is dropped.

Is it the normal behaviour ? Or is it a bug ?

### PR DESCRIPTION
**Pull request by tienvx**
| Q             | A
| ------------- | ---
| Branch?       | 4.3 
| Bug fix?      | yes
| New feature?  | no 
| Deprecations? | no 
| Tickets       | Fix #32311
| License       | MIT
| Doc PR        | NA

### CODE REVIEW

**Comment by tienvx** [comment]
**NOTE**: The test case will failed on Symfony 4.4, because component `Debug` is removed from dependencies of the component `Messenger`

File: src/Symfony/Component/Messenger/Tests/Command/FailedMessagesRetryCommandTest.php [hunk 1]
```diff
@@ -11,14 +11,18 @@
 
 namespace Symfony\Component\Messenger\Tests\Command;
 
+use Exception;
```
Comment (fabpot) [inline src/Symfony/Component/Messenger/Tests/Command/FailedMessagesRetryCommandTest.php]: We don't use `use` for PHP built-on classes. Can you use `\Exception` in the code instead?

File: src/Symfony/Component/Messenger/Tests/Command/FailedMessagesRetryCommandTest.php [hunk 2]
```diff
@@ -46,4 +50,52 @@ public function testBasicRun()
...
+        $retryStrategy = $this->createMock(RetryStrategyInterface::class);
+        $retryStrategy->expects($this->once())->method('isRetryable')->with($this->isInstanceOf(Envelope::class))->willReturn(true);
+
+        $command = new FailedMessagesRetryCommand(
+            'failure_receiver',
+            $receiver,
+            $bus,
+            $dispatcher,
+            $retryStrategy
+        );
```
Comment (fabpot) [inline src/Symfony/Component/Messenger/Tests/Command/FailedMessagesRetryCommandTest.php]: Could be on one line instead.

File: src/Symfony/Component/Messenger/Tests/Command/FailedMessagesRetryCommandTest.php [hunk 3]
```diff
@@ -46,4 +50,52 @@ public function testBasicRun()
...
+        $dispatcher = $this->createMock(EventDispatcherInterface::class);
+        $bus = $this->createMock(MessageBusInterface::class);
+        // the bus should be called in the worker
+        $bus->expects($this->at(0))
+            ->method('dispatch')
+            ->with($this->callback(function (Envelope $envelope) {
+                $lastReceivedStamp = $envelope->last(ReceivedStamp::class);
+
+                return $lastReceivedStamp instanceof ReceivedStamp &&
+                    \is_string($lastReceivedStamp->getTransportName());
```
Comment (fabpot) [inline src/Symfony/Component/Messenger/Tests/Command/FailedMessagesRetryCommandTest.php]: Could be on one line

File: src/Symfony/Component/Messenger/Worker.php [hunk 4]
```diff
@@ -215,4 +216,17 @@ private function shouldRetry(\Throwable $e, Envelope $envelope, RetryStrategyInt
...
     }
+
+    private function flattenedException(\Throwable $throwable): ?FlattenException
+    {
+        if (class_exists(FlattenException::class)) {
+            if ($throwable instanceof HandlerFailedException) {
+                return FlattenException::createFromThrowable($throwable->getNestedExceptions()[0]);
+            } else {
+                return FlattenException::createFromThrowable($throwable);
+            }
```
Comment (fabpot) [inline src/Symfony/Component/Messenger/Worker.php]: To respect our CS, this should  be rewritten as:

```php
if (!class_exists(FlattenException::class)) {
    return null;
}

if ($throwable instanceof HandlerFailedException) {
    $throwable = $throwable->getNestedExceptions()[0];
}

return FlattenException::createFromThrowable($throwable);
```

**Review by fabpot — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Messenger/Tests/Command/FailedMessagesRetryCommandTest.php [hunk 5]
```diff
@@ -46,4 +50,52 @@ public function testBasicRun()
...
+            ->will($this->throwException(new Exception('Mock test exception')));
+
+        $bus->expects($this->at(1))
+            ->method('dispatch')
+            ->with($this->callback(function (Envelope $envelope) {
+                $lastRedeliveryStamp = $envelope->last(RedeliveryStamp::class);
+
+                return $lastRedeliveryStamp instanceof RedeliveryStamp &&
+                    \is_string($lastRedeliveryStamp->getExceptionMessage()) &&
+                    $lastRedeliveryStamp->getFlattenException() instanceof FlattenException;
```
Comment (tienvx) [inline src/Symfony/Component/Messenger/Tests/Command/FailedMessagesRetryCommandTest.php]: @fabpot Do you want me to update this to be one line too?

File: src/Symfony/Component/Messenger/Tests/Command/FailedMessagesRetryCommandTest.php (hunk 2 above)
Comment (tienvx) [reply src/Symfony/Component/Messenger/Tests/Command/FailedMessagesRetryCommandTest.php]: Do you want me to update the same for `testBasicRun` test case too?

File: src/Symfony/Component/Messenger/Tests/Command/FailedMessagesRetryCommandTest.php (hunk 5 above)
Comment (fabpot) [reply src/Symfony/Component/Messenger/Tests/Command/FailedMessagesRetryCommandTest.php]: This one looks good to me.

File: src/Symfony/Component/Messenger/Tests/Command/FailedMessagesRetryCommandTest.php (hunk 1 above)
Comment (tienvx) [reply src/Symfony/Component/Messenger/Tests/Command/FailedMessagesRetryCommandTest.php]: Thanks for reviewing!

**Review by fabpot — APPROVED** [review]

**Comment by Tobion** [comment]
This fixes #32311 and invalidates #32341 where @weaverryan wanted to go with a different solution.

**Comment by TimoBakx** [comment]
This most likely also invalidates #32904, unless the separation is still something we want to do in the future.

**Comment by tienvx** [comment]
Thanks @Tobion and @TimoBakx for adding references

**Comment by fabpot** [comment]
I wasn't aware of the other issues/PRs; tell me how you want to proceed @weaverryan

**Comment by Tobion** [comment]
We need to revert this in #34082 as it's breaking the retry on amqp when the message gets too big.

**Pull request merged**